rskj-core/src/test/java/org/ethereum/datasource/DataSourceWithCacheTest.java

```diff
@@ -130,6 +157,51 @@ public void putTwoKeyValuesWrittenInOrder() {
         order.verify(baseDataSource).put(randomKey2, randomValue2);
     }
 
+    @Test
+    public void putLockWorks() {
+        ReentrantReadWriteLock lock = TestUtils.getInternalState(dataSourceWithCache, "lock");
+        lock.writeLock().lock();
+        boolean unlocked = false;
+
+        try {
+            Map<ByteArrayWrapper, byte[]> committedCache = spy(new HashMap<>());
+            TestUtils.setInternalState(dataSourceWithCache, "committedCache", committedCache);
+
+            byte[] randomKey1 = TestUtils.randomBytes(20);
+            ByteArrayWrapper randomKeyWrapped = ByteUtil.wrap(randomKey1);
+            byte[] randomValue1 = TestUtils.randomBytes(20);
+            byte[] randomValueAfterLock = TestUtils.randomBytes(20);
+            committedCache.put(randomKeyWrapped, randomValue1); // to check how many times remove is called
+
+            AtomicBoolean threadStarted = new AtomicBoolean(false);
+            ExecutorService executor = Executors.newSingleThreadExecutor();
+            Future<?> future = executor.submit(new Thread(() -> {
+                threadStarted.set(true);
+                dataSourceWithCache.put(randomKey1, randomValueAfterLock);
+            }));
+
+            // wait for thread to be started and put a value while thread is locked
+            Awaitility.await().timeout(Duration.ofMillis(100)).pollDelay(Duration.ofMillis(10)).untilAtomic(threadStarted, equalTo(true));
```

Comment:
isn't `100ms`  too short timeout? tests may fail due to system being overloaded, what do you think?

Reply:
yeah, you are right, even if it is waiting just for the thread to start, not to finish, it wouldn't harm if we increase that value (being this a test and having such short polls) and this may prevent some edge scenarios where it is not possible create more threads or whatever 👍